### DISCUSSION TITLE
[Proposal] A cache line aware new operator

### DISCUSSION THREAD

**Discussion by gaw79** [dotnet/csharplang#5956] [Language Ideas]
Greetings.

Add a special version of the new operator which guarantees that the allocated object will never share a memory cache line (or cache lines) with any other object.  The purpose is to selectively prevent the possibility of false sharing from occurring with certain objects.  This could bring significant speed improvements to some multi-threaded code.

For the discussion, I will call this operator "cnew", though the actual syntax is of low importance.

When a block of memory of size N bytes is allocated via cnew, it is aligned at the start of a cache line, and padded with extra bytes, if necessary, so that it ends on a cache line.  The actual amount of memory allocated is (floor(N/C)+((N%C)>0?1:0))*C where C is the cache line size.  If there are multiple cache levels with different cache line sizes, then C is the least common multiple of the different cache line sizes.

A block of memory allocated with cnew is called aligned. A block of memory allocated without the cache line alignment and padding (the old way) is called unaligned.

If an aligned block is pinned, it will still be aligned after the pin.

Additionally, there would a library call (perhaps GC.SetAlignMode()) to control all new/cnew operations:
- AlignMode.AsIs - subsequent blocks allocated with cnew are aligned, and blocks allocated with new are unaligned.
- AlignMode.AlwaysAligned - all subsequent memory blocks are aligned, regardless of whether new or cnew was used.
- AlignMode.AlwaysUnaligned - all subsequent memory blocks are unaligned, regardless of whether new or cnew was used.

This library call would make it easier to test for the existence of performance-affecting false sharing issues.

The drawbacks and unresolved issues are:
1. Is it even possible for the .NET memory allocation subsystem to accurately determine the cache line sizes for cache levels at run time?
2. The memory allocation system needs to keep track of the aligned/unaligned state of each memory block.  The details of this would have to be worked out, but it would be additional overhead.
3. Should there be a method to query the aligned/unaligned state of an object?  Is there a use case for this functionality?
4. Should there be a method to change the aligned/unaligned state of an object?  Is there a use case for this functionality?

As for syntax, it would be new followed by an optional keyword. For example:
`SomeClass p = new aligned SomeClass();`

Thanks.

**Comment by huoyaoyuan** [comment]
Like `GC.AllocateUninitializedArray`, this shouldn't be an operator of language. It should be a method exposed by the runtime.

**Reply by jnm2** [reply]
I cnew someone would point this out.

**Reply by gaw79** [reply]
In response to huoyaoyuan's comment, I suggest that a new keyword **_not_** be introduced.  Instead, add new versions of the `Activator.CreateInstance()` method which include a new parameter to specify the align mode. For example:

`Activator.CreateInstance(Type, Object[], ObjectAlignMode)`

Where `ObjectAlignMode` is an enum with two initial values: `NotAligned`, and `CacheAligned`. An enum is chosen instead of a bool to allow for other future alignment options.

The existing versions of` CreateInstance()` without an `ObjectAlignMode` parameter default to `NotAligned`.

I think this change removes the necessity of the `GC.SetAlignMode()` method.

**Reply by huoyaoyuan** [reply]
`Activator.CreateInstance()` is way less efficient than generic factory method.
